portal-impl/src/com/liferay/portal/internal/servlet/MainServlet.java

```diff
@@ -788,19 +789,25 @@ private void _initCompanies() throws Exception {
 			throw new RuntimeException("Company default web ID is null");
 		}
 
+		List<Company> companies = new ArrayList<>();
+
 		CompanyLocalServiceUtil.forEachCompany(
```

Comment:
And please get @tinatian involved for this PortalInstancePool change. She created that class.

Reply:
Hi @shuyangzhou,

That won't happen because of this flag:
https://github.com/shuyangzhou/liferay-portal/blob/3468ce7f87f714e0e22a6689639c8dd0079cc9be/portal-impl/src/com/liferay/portal/util/PortalInstances.java#L321

In the _forEachCompany_ the flag is true so we won't add the company to the pool during the _initCompany_.

Anyways, we agree the flag is not the ideal solution so we added this instead:
https://github.com/shuyangzhou/liferay-portal/blob/3468ce7f87f714e0e22a6689639c8dd0079cc9be/portal-impl/src/com/liferay/portal/util/PortalInstances.java#L321

Because adding a company individually does not make sense until the company map is ready.

Regarding splitting this pull, could you finish your review, now it is in progress, and then we can separate it so @tinatian can check the PortalInstancePool changes if needed? This has been under review for several weeks and start the process again will impact our planning with LXC.

Thanks.

Reply:
I don't have much problems with the rest of the pull. All issues are around PortalInstancePool. Please go through Tina with the PortalInstancePool change.

Reply:
Hi @shuyangzhou, thank you for the review! I am closing the PR and going through Tina with the PortalInstancePool change.